src/Sema.zig

```diff
@@ -15446,6 +15629,56 @@ fn wantErrorReturnTracing(sema: *Sema) bool {
         backend_supports_error_return_tracing;
 }
 
+fn zirSaveErrRetIndex(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.Inst.Ref {
+    const inst_data = sema.code.instructions.items(.data)[inst].node;
+    const src = LazySrcLoc.nodeOffset(inst_data);
+
+    // This is only relevant at runtime.
+    if (block.is_comptime) return Air.Inst.Ref.zero_usize;
+
+    // In the corner case that `catch { ... }` or `else |err| { ... }` is used in a function
+    // that does *not* make any errorable calls, we still need an error trace to interact with
+    // the AIR instructions we've already emitted.
+    if (sema.owner_func != null)
+        sema.owner_func.?.calls_or_awaits_errorable_fn = true;
```

Comment:
Do you have an example of this case? Also `sema.owner_func` is guaranteed non-null here since we already checked that the block isn't comptime.

Reply:
Good tip.

Here's an example:
```zig
fn non_errorable() void {
    // Make sure catch works even in a function that does not call any errorable functions.
    //
    // This test is needed because stage 2's fix for #1923 means that catch blocks interact
    // with the error return trace index.
    var x: error{Foo}!void = {};
    return x catch {};
}

test "catch within a function that calls no errorable functions" {
    non_errorable();
}
```

Intended to add this as a test, but I accidentally left it behind in a stash - Nice catch.

Reply:
And why does that need to save the error return trace index? There are no errors being returned anywhere in that example?

Reply:
The problem is that in "foo() catch { ... }" the `.save_err_ret_index` precedes the error-able call, so Sema does not know that an errorable call is made or not until after we analyze that `.save_err_ret_index` instruction. 

A similar problem means that `@errorReturnTrace()` allows returns `null` in a function before the first errorable function call, even if that function does call errorable functions and has a live error trace:
```zig
fn bar() !void { return error.TheSkyIsFalling; }

fn foo() !void {
    // "error: unable to unwrap null" if uncommented
    //std.log.err("entry: {}", .{ @errorReturnTrace().?.index });

    bar() catch {
        std.log.err("in catch: {}", .{ @errorReturnTrace().?.index });
    };

    std.log.err("exit: {}", .{ @errorReturnTrace().?.index });
}
```

Reply:
I think the behavior with `@errorReturnTrace` is desired. And wouldn't every `.save_err_ret_index` before a call to an erorrable function return zero anyways?

Reply:
Not in the case that `foo()` is being called from within a `catch` block, for example

Reply:
But `foo` doesn't return any errors either?

Reply:
Sorry, I meant for a general `foo()` not the example above

Reply:
How could the error return trace be anything but zero if it is inspected before any operation that could set it to a non-zero value?

Reply:
The index can be updated by the caller before the function entry. This is how error traces are chained together:
```zig
fn foo() !void { return error.BigIssue; }
fn bar() !void { return error.AnotherProblem; }
fn baz() !void {
    // error trace is not empty at function entry, since we were called from within a `catch`
    return bar();
}
test {
    foo() catch {
        // @errorReturnTrace.?.index == 1 here
        baz() catch unreachable;
    }
}
```

Reply:
Sounds good to me. I'll change to that fix in my next round of edits :+1: